numpyro/distributions/distribution.py

```diff
@@ -193,6 +193,16 @@ def event_dim(self):
         """
         return len(self.event_shape)
 
+    @property
+    def has_rsample(self):
+        return not self.is_discrete
```

Comment:
We already have a more granular `reparametrized_params` available. I'm not sure if all our continuous distributions are reparametrized? If so, a safer choice might be `return set(self.reparametrized_params) == set(self.arg_constraints.keys())`.

Reply:
Makes sense to me. I'll make the change. One thing is reparameterized notion is a bit ambiguous. If I understand correctly, TFP folks implement reparameterized sampler for gamma, then the paper implies dirichlet or studentt using gamma sampler are also reparameterized. But it is a bit different from @fritzo and @martinjankowiak work, where any derived sampler needs its own reparam sampler.

Reply:
fyi: I would say that also in Martin's and my work, we get downstream samplers like Dirichlet and StudentT for free. However those "free" samplers used by tfp involve latent variables (e.g. the gamma in a student t sampler), and therefore introduce variance. What Martin and I did was to derive a custom reparametrization for Dirichlet that is deterministic and therefore has lower variance than the "free" version that tfp uses. But in both cases it is safe to set `.has_rsample = True`.

Reply:
@fehiepsi - I wasn't sure about distributions like `LKJCholesky`, for instance.

Reply:
@fritzo Thanks! I was hesitating to set `has_rsample=True` for Dirichlet/StudentT because I couldn't make them pass the pathwise tests. But it seems that the "free" version is not expected to pass those tests because they have higher variance. I'll update the PR to set `has_rsample=True` for those distributions.

@neerajprad Looking like we can set `rsample` for it because it is a transformed distribution of a Beta distribution. :D